### PR TITLE
Improve comparison of sensitive marks on resources, and propagate the sensitive_attributes correctly

### LINKED ISSUE #1616: Migration to tofu1.7.0 from terraform1.8.2 shows updated in-place required for resources even though nothing has changed
### OpenTofu Version

```shell
terraform -v
Terraform v1.8.2
on windows_amd64
+ provider registry.terraform.io/hashicorp/azurerm v3.85.0
+ provider registry.terraform.io/hashicorp/random v3.6.1

tofu -v
OpenTofu v1.7.0
on windows_amd64
+ provider registry.opentofu.org/hashicorp/azurerm v3.85.0
+ provider registry.opentofu.org/hashicorp/random v3.6.1
```

### OpenTofu Configuration Files

```hcl
terraform {
  required_providers {
    azurerm = {
      source = "hashicorp/azurerm"
      version = "=3.85.0"
    }
  }
}

# 2. Configure the AzureRM Provider
provider "azurerm" {
  skip_provider_registration = true
  features {}
}

data "azurerm_resource_group" "tofu-rg" {
  name     = "testrg"

}

data "azurerm_key_vault" "tofu-kv" {
  name                = "tofu-kv"
  resource_group_name = data.azurerm_resource_group.tofu-rg.name
}

resource "azurerm_key_vault_secret" "sample1" {
  name         = "SAMPLE1"
  value        = random_password.password1.result
  key_vault_id = data.azurerm_key_vault.tofu-kv.id
}

resource "random_password" "password1" {
  length           = 12
  special          = false
}
```

### Debug Output

https://gist.github.com/mccoyer001/2f16280ecff58252edd4f25b4d2d2b24

### Expected Behavior

"tofu plan" should not detect any changes required since the "terraform plan" shows no changes detected.

### Actual Behavior

tofu plan shows the need for updated in-place

### Steps to Reproduce

# initialize the terraform env and apply configuration and generate local terraform.tfstate file 
1. terraform init
2. terraform plan
3. terraform apply
# cleanup the terraform registry files 
4. del .\.terraform
5. del .\.terraform.lock.hcl
# initialize the tofu env and validate config has not changed
6. tofu init
7. tofu plan

### Additional Context

Have also seen this "updated in-place" false reporting for azure key vault secrets in my other setups. 

### References

_No response_

### PR DESCRIPTION
**Pull request by Yantrio**
This PR resolves 2 things primarily.

- Ensure that we propagate the sensitive_attributes correctly to the statefile. This was raised as a possible point of concern by @cam72cam [here](https://github.com/opentofu/opentofu/issues/1616#issuecomment-2097282091)
- Move the mark comparison logic to just use the prior and the proposed values when detecting if a NoOp change is actually a change of sensitivity.

It also introduced a method to combine `[]cty.PathValueMark` which is used to ensure that `sensitive_attributes` are not being thrown away. 

This is used to combine the existing value marks with the marks from the configuration. 
Because we handle it this way now, the sensitive_attributes value in the statefile will now include all sensitive properties and not just those where a sensitive value was passed in.

Resolves #1616

## Target Release
1.8.0

### CODE REVIEW

**Comment by tun0** [comment]
Out of curiosity, what's the status of/planning for this PR?

**Comment by Yantrio** [comment]
Hi @tun0 it looks like some of the snapshot tests need updating. I'll attempt to do this now so we can get this into the next release.

File: internal/tofu/marks.go [hunk 1]
```diff
@@ -60,6 +44,7 @@ func marksEqual(a, b []cty.PathValueMarks) bool {
 }
 
 func copyMarksFromValue(dst, src cty.Value) cty.Value {
+	// TODO JAMES, DONT LET ME MERGE THIS PR WITH THIS IN!!: Rework this to merge marks, not just copy
```
Comment (cam72cam) [inline internal/tofu/marks.go:47]: Hey James, don't forget about this :P

Comment (Yantrio) [reply internal/tofu/marks.go]: Fixed!

File: internal/tofu/node_resource_abstract_instance.go [hunk 2]
```diff
@@ -662,14 +662,64 @@ func (n *NodeAbstractResourceInstance) refresh(ctx EvalContext, deposedKey state
...
 	}
 
 	return ret, diags
 }
 
+// combinePathValueMarks will combine the marks from two sets of marks with paths, ensuring that we don't duplicate marks
+// for the same path, but instead combine the marks for the same path
+// This ensures that we don't lose user marks when combining 2 different sets of marks for the same path
+func combinePathValueMarks(marks []cty.PathValueMarks, other []cty.PathValueMarks) []cty.PathValueMarks {
+	// TODO: Find a nice home for this method so it can be used in other places too
```
Comment (cam72cam) [inline internal/tofu/node_resource_abstract_instance.go:682]: Should this live in internal/tofu/marks.go?

Comment (Yantrio) [reply internal/tofu/node_resource_abstract_instance.go:682]: Yes, I'll move that now :D

Comment (Yantrio) [reply internal/tofu/node_resource_abstract_instance.go]: [5b0a7f8](https://github.com/opentofu/opentofu/pull/1640/commits/5b0a7f8ad7f1ede6e2b670b34a3f7e68d53d9080)

**Review by cam72cam — DISMISSED** [review]

File: internal/tofu/marks.go [hunk 3]
```diff
@@ -59,10 +43,45 @@ func marksEqual(a, b []cty.PathValueMarks) bool {
...
+		for i, existing := range combined {
+			if mark.Path.Equals(existing.Path) {
+				// if we found a matching path, we should combine the marks and update the existing item
+				dupe := existing
+				for k, v := range mark.Marks {
+					dupe.Marks[k] = v
+				}
+				combined[i] = dupe
+				exists = true
+				break
```
Comment (ollevche) [inline internal/tofu/marks.go:76]: Are we sure `marks` slice contains unique paths?

**Review by ollevche — COMMENTED** [review]
Looks nice overall, thanks! I left a couple of questions to understand the change better

File: internal/tofu/marks.go [hunk 4]
```diff
@@ -59,10 +43,45 @@ func marksEqual(a, b []cty.PathValueMarks) bool {
...
+
+	// check if we've already inserted this by looping over and calling .Equals().
+	// This isn't so nice but there is no nice comparison for cty.PathValueMarks
+	// so we have to do it this way
+	for _, mark := range other {
+		exists := false
+		for i, existing := range combined {
+			if mark.Path.Equals(existing.Path) {
+				// if we found a matching path, we should combine the marks and update the existing item
+				dupe := existing
```
[the remark this answers is not in the document]
Comment (Yantrio) [reply internal/tofu/marks.go:70]: yes we should 👍

File: internal/tofu/marks.go (hunk 3 above)
Comment (Yantrio) [reply internal/tofu/marks.go:76]: it's a map under the hood, so in theory it should be unique 👍

**Review by ollevche — APPROVED** [review]

**Review by cam72cam — APPROVED** [review]

**Pull request merged**